graphstorm-processing/graphstorm_processing/graph_loaders/dist_heterogeneous_loader.py

```diff
@@ -0,0 +1,1560 @@
+                        logging.info(
+                            "Re-encountered node type '%s' in edge file for "
+                            "'%s:%s:%s', "
+                            "re-creating mapping for it using '%s' as join column...",
+                            ntype,
+                            src_type,
+                            rel_type,
+                            dst_type,
+                            ncol,
+                        )
+                        node_type_to_mapping_df[ntype] = self._extend_mapping_from_edges(
+                            mapping_df, edges_df, ncol
+                        )
+                    else:
+                        node_type_to_mapping_df[ntype] = self._create_initial_mapping_from_edges(
+                            edges_df, ncol
+                        )
+        # Write back each mapping and populate self.node_mapping_paths
+        for ntype, mapping_df in node_type_to_mapping_df.items():
+            # TODO: This check is probably redundant
+            if ntype in missing_node_types:
+                self._write_nodeid_mapping_and_update_state(mapping_df, ntype)
+
+    def _create_initial_mapping_from_edges(self, edges_df: DataFrame, node_col: str) -> DataFrame:
+        assert isinstance(node_col, str), f"{node_col=} not of str type, got {type(node_col)=}"
+        distinct_nodes_df = edges_df.select(node_col).distinct()
+        # TODO: Instead of using the below that triggers a computation,
+        # skip zipWithIndex and only trigger at write time?
+        node_mapping_df = self.create_node_id_map_from_nodes_df(distinct_nodes_df, node_col)
+
+        return node_mapping_df
+
+    def _extend_mapping_from_edges(
+        self, mapping_df: DataFrame, edges_df: DataFrame, node_col: str
+    ) -> DataFrame:
+        """Extends existing node mapping from incoming edge structure.
+
+        Given an existing node ud mapping df, incoming edges_df and a node column,
+        will extend the mapping to include any new node ids from the
+        incoming edges DataFrame.
+
+        Parameters
+        ----------
+        mapping_df : DataFrame
+            An existing mapping from node string ids to numerical ids.
+        edges_df : DataFrame
+            An incoming edges structure that includes the node type.
+        node_col : str
+            The name of the column that corresponds to the node type
+            in the incoming edges DataFrame.
+
+        Returns
+        -------
+        DataFrame
+            Returns a new mapping DataFrame, ordered by the updated
+            integer node ids.
+        """
+        join_col = NODE_MAPPING_STR
+        index_col = NODE_MAPPING_INT
+        incoming_node_ids = (
+            edges_df.select(node_col).distinct().withColumnRenamed(node_col, join_col)
+        )
+
+        # TODO: When is it possible that we might get a null node str id value?
+        # Is it only in the case the actual row has null values?
```

Comment:
Since we won't be able to join on those one option is to remove all nulls at the node level, e.g. gr node ids through `edges_df.select(node_col).distinct().dropna()`. 

Would that suffice?

Reply:
Can we count the number of rows in edges_df and number of rows after dropna()?

Reply:
Unless there's good reason to do so, I'd like to avoid calling `count()` as it triggers a computation as mentioned elsewhere. If you think it's necessary I can add the check though.

What do we want to do in case there's null though? Do we want to print a warning and move on?

Reply:
Added extra assertion, resolving.